### PR TITLE
Editable function test messages

### LINKED ISSUE #1550: Enable inline editing of test messages in the Functions editor
### **Background / Problem**

When creating or editing a function in Sequin, we let users work with test messages. That test message is either:

* Our synthetic (hard-coded) message (default)
* We also let users select a database + table and capture test messages flowing from that table for them to use in the editor

Today, both of these test messages are read-only.<br>Many users want to tweak the sample record, change set, action, or metadata to reproduce edge-cases or scenarios that are hard to generate in their database.

---

### **Goal**

Let users edit every part of the test message directly in the Functions UI so they can iterate on transforms/routing faster.

---

### **Scope (MVP)**

| Field | Editing UX | Notes |
| -- | -- | -- |
| **record** (Elixir map) | “✏️ Edit” link/icon → converts to a CodeMirror editor pre-populated with the map | Users supply valid Elixir; no schema validation in MVP |
| **changes** (Elixir map) | Same as **record** | — |
| **action** | Dropdown with the 4 allowed atoms: `:insert`, `:update`, `:delete`, `:read` | `:read` is used for backfills |
| **metadata** (map) | Same editable CodeMirror block | Keys are fixed today but remain freely editable in MVP |

*The edited message is passed to the function whenever the user clicks **Run Test**.*

Persisting custom test messages (CRUD) and sharing them across sessions is **out of scope** for this ticket.

Note that you'll need to use our MiniElixir validator to ensure safe eval/conversion of test messages into an Elixir map. (It's OK if it's not perfect — we probably should error if they give us something that is not `is_map`!)

---

### **Nice-to-have / Stretch**

* Save the last-edited message to `localStorage` so a refresh preserves the user’s work (do **not** block MVP).
* Disable **Run Test** until the edited JSON/Elixir parses successfully to prevent runtime errors.

---

### **Acceptance Criteria**

1. **Edit Controls**
   * Pencil icon or “Edit” link shown next to **record**, **changes**, **action**, and **metadata**; dropdown on edit for **action**, CodeMirror for everything else.
2. **Editor**
   * Clicking “Edit” toggles a syntax-highlighted CodeMirror textarea (Elixir mode) in place.
   * “Save” button (or blur) returns to the read-only view.
   * “Cancel” reverts changes.
   * (These could be either icons or link words in the top right)
3. **Validation**
   * Basic parse check; show inline error if the Elixir map cannot be parsed.
4. **Execution**
   * Pressing **Run Test** invokes the function with the edited message.
5. **No Persistence**
   * Reloading the page resets the test message (unless stretch goal implemented).
6. **No backend CRUD**
   * Feature ships without storing or listing saved test messages.

---

### **Out of Scope**

* Backend storage (CRUD) of custom test messages.
* Advanced metadata editor that locks keys in a table view.

### PR DESCRIPTION
**Pull request by davoclavo**
### Implement editable attributes in function test messages, both synthetic and real test messages

A pencil is now shown at the right of each attribute that will be passed to the function. If the user clicks on the pencil it transforms into a 
Save and Cancel icons
- For "record", "metadata", "changes" arguments, it converts the text into a CodeMirror instance.
- For the "action" argument, a dropdown is shown instead

Save / Cancel logic:
- If the cancel button is pressed the text is left unchanged and resets the editor/dropdown back to the original value.

- If the save button is pressed, the encoded version is updated in the svelte component, which causes the validation event to occur. Within the validation, the encoded changes are converted into structs, and are injected into the socket assigns for the synthetic and real test messages.

Side notes:
- I did not add a Run Test button as I think its simpler to just run tests on each save, for a quicker feedback. Let me know if this behavior is not desired and an additional button should be added.
- I am aware of the current rename of transforms -> functions, I am prepared to rebase if needed.
- Implemented a hacky way for minielixir to run interpreted path functions

Any feedback is more than welcome!

Fixes #1550

---

### demo

https://github.com/user-attachments/assets/ce939daf-83f7-493d-991f-7629ca613af8

### CODE REVIEW

File: lib/sequin/functions/minielixir.ex [hunk 1]
```diff
@@ -236,6 +248,14 @@ defmodule Sequin.Functions.MiniElixir do
     %{type: "Unknown error", info: %{description: Exception.message(error)}}
   end
 
+  def encode_error({error_info, message, _rest}) when is_list(error_info) do
+    %{type: "Syntax error", info: %{description: message}}
```
Comment (davoclavo) [inline lib/sequin/functions/minielixir.ex:252]: I should add a couple of tests for these new variants of encode_error

File: assets/svelte/functions/EditableArgument.svelte [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+
+  interface TestMessage {
+    id: string;
+    record: string;
+    changes: string;
+    action: ActionType;
+    metadata: string;
+  }
+
+  export let form;
```
Comment (davoclavo) [inline assets/svelte/functions/EditableArgument.svelte:28]: Add type annotations for `form` and `formErrors`

File: assets/svelte/functions/EditableArgument.svelte [hunk 3]
```diff
@@ -0,0 +1,169 @@
+<script lang="ts">
```
Comment (davoclavo) [inline assets/svelte/functions/EditableArgument.svelte:1]: Is there a convention for private svelte components? For example this one is strongly tied to the form data in Edit.svelte

File: lib/sequin/functions/minielixir.ex [hunk 4]
```diff
@@ -66,7 +66,19 @@ defmodule Sequin.Functions.MiniElixir do
       metadata: Sequin.Map.from_struct_deep(data.metadata)
     ]
 
-    outer = Code.string_to_quoted!(code)
+    outer =
```
Comment (davoclavo) [inline lib/sequin/functions/minielixir.ex:69]: Should add tests for path functions

**Comment by RTLS** [comment]
This looks so good! Still need to review the PR but the video looks awesome 👏 @davoclavo

File: assets/svelte/functions/Edit.svelte [hunk 5]
```diff
@@ -1050,52 +1132,61 @@ Please help me create or modify the Elixir function transform to achieve the des
           </div>
           <div class="p-4 space-y-2">
             {#each messagesToShow as message, i}
-              <button
-                class="w-full text-left p-3 rounded-lg border border-slate-200 dark:border-slate-800 transition-all duration-200 {selectedMessageIndex ===
+              <div
```
Comment (fazzone) [inline assets/svelte/functions/Edit.svelte:1135]: 👌

File: assets/svelte/functions/Edit.svelte [hunk 6]
```diff
@@ -1050,52 +1132,61 @@ Please help me create or modify the Elixir function transform to achieve the des
...
-                  <span class="font-medium"
-                    >{showSyntheticMessages
-                      ? "Example"
-                      : `Message ${i + 1}`}</span
-                  >
-                </div>
+                <button
+                  class="w-full p-3"
+                  on:click={() => selectMessage(i)}
+                  type="button"
```
Comment (fazzone) [inline assets/svelte/functions/Edit.svelte:1144]: 👌

File: lib/sequin_web/live/functions/edit.ex [hunk 7]
```diff
@@ -186,12 +213,109 @@ defmodule SequinWeb.FunctionsLive.Edit do
...
+         {:ok, action} <- validate_action(message["action"]),
+         {:ok, metadata} <- validate_elixir_code(message["metadata"], "metadata"),
+         {:ok, changes} <- validate_elixir_code(message["changes"], "changes") do
+      {:ok, %{record: record, action: action, metadata: metadata, changes: changes}}
+    else
+      {:error, field, error} -> %{error: %{field => error}}
+    end
+  end
+
+  defp validate_elixir_code(code, field) do
```
Comment (fazzone) [inline lib/sequin_web/live/functions/edit.ex:282]: I think this can live in minielixir module

File: lib/sequin_web/live/functions/edit.ex [hunk 8]
```diff
@@ -186,12 +213,109 @@ defmodule SequinWeb.FunctionsLive.Edit do
...
+            {id, validate_test_message(message)}
+          end)
+      end
+
+    synthetic_test_message = socket.assigns.synthetic_test_message
+
+    new_synthetic_test_message =
+      case modified_test_messages[synthetic_test_message.id] do
+        {:ok, result} ->
+          %{synthetic_test_message | data: update_message_data(synthetic_test_message.data, result)}
```
Comment (fazzone) [inline lib/sequin_web/live/functions/edit.ex:230]:
```suggestion
          %{synthetic_test_message | data: Map.merge(synthetic_test_message.data, result)}
```

File: lib/sequin_web/live/functions/edit.ex [hunk 9]
```diff
@@ -186,12 +213,109 @@ defmodule SequinWeb.FunctionsLive.Edit do
...
+        _ ->
+          synthetic_test_message
+      end
+
+    # Update regular test messages
+    new_test_messages =
+      Enum.map(socket.assigns.test_messages, fn message ->
+        case modified_test_messages[message.id] do
+          {:ok, result} ->
+            %{message | data: update_message_data(message.data, result)}
```
Comment (fazzone) [inline lib/sequin_web/live/functions/edit.ex:241]:
```suggestion
            %{message | data: Map.merge(message.data, result)}
```

File: lib/sequin_web/live/functions/edit.ex [hunk 10]
```diff
@@ -186,12 +213,109 @@ defmodule SequinWeb.FunctionsLive.Edit do
...
+      case params["modified_test_messages"] do
+        messages when is_map(messages) ->
+          Map.new(messages, fn {id, message} ->
+            {id, validate_test_message(message)}
+          end)
+      end
+
+    synthetic_test_message = socket.assigns.synthetic_test_message
+
+    new_synthetic_test_message =
```
Comment (fazzone) [inline lib/sequin_web/live/functions/edit.ex:227]: We love shadowing and routinely write e.g.
```
a = initial
a = f(a)
```
this is not a suggestion, it's up to you!

File: lib/sequin_web/live/functions/edit.ex [hunk 11]
```diff
@@ -186,12 +213,109 @@ defmodule SequinWeb.FunctionsLive.Edit do
...
+        e -> {:error, field, MiniElixir.encode_error(e)}
+      end
+    else
+      {:error, error_type, error} -> {:error, field, MiniElixir.encode_error(error_type, error)}
+      {:error, error} -> {:error, field, MiniElixir.encode_error(error)}
+    end
+  end
+
+  defp validate_action(action) when is_binary(action) do
+    case action |> String.trim("\"") |> String.downcase() do
```
Comment (fazzone) [inline lib/sequin_web/live/functions/edit.ex:298]:
```suggestion
    case action |> String.trim("\" ") |> String.downcase() do
```

File: lib/sequin/functions/minielixir.ex [hunk 12]
```diff
@@ -66,7 +66,19 @@ defmodule Sequin.Functions.MiniElixir do
...
+          function.path
+          |> String.split(".")
+          |> Enum.reduce(fn part, acc ->
+            "#{acc}[\"#{part}\"]"
+          end)
+
+        Code.string_to_quoted!("def transform(action, record, changes, metadata), do: #{path}")
+      else
+        Code.string_to_quoted!(function.code)
+      end
```
Comment (fazzone) [inline lib/sequin/functions/minielixir.ex:81]: consider:
```
unwrapped =
  if function.type == :path do
    parts = String.split(function.path, ".")
    {:ok, quote(do: get_in(record, unquote(parts)))}
  else
    function.code
    |> Code.string_to_quoted!()
    |> Validator.unwrap()
  end
```

File: assets/svelte/functions/Edit.svelte [hunk 13]
```diff
@@ -376,6 +375,78 @@
...
+      messageRecordEditorView = null;
+    }
+
+    const messageRecordEditorState = EditorState.create({
+      doc: message.record,
+      extensions: [
+        basicSetup,
+        elixir(),
+        autocompletion({ override: [] }),
+        keymap.of([indentWithTab]),
```
Comment (fazzone) [inline assets/svelte/functions/Edit.svelte:440]: ctrl/cmd-enter would be 👌

**Review by fazzone — APPROVED** [review]
very nice!  The idea with path functions is to just to elixir-ify them so there can be one code path?

File: lib/sequin/functions/minielixir.ex (hunk 12 above)
Comment (davoclavo) [reply lib/sequin/functions/minielixir.ex:81]: ahh nice!

File: lib/sequin_web/live/functions/edit.ex (hunk 8 above)
Comment (davoclavo) [reply lib/sequin_web/live/functions/edit.ex]: ooof thanks!! 😅

File: lib/sequin_web/live/functions/edit.ex (hunk 10 above)
Comment (davoclavo) [reply lib/sequin_web/live/functions/edit.ex]: thanks, sounds good to me!

File: lib/sequin_web/live/functions/edit.ex (hunk 11 above)
Comment (davoclavo) [reply lib/sequin_web/live/functions/edit.ex]: These specific `action` values are coming from the dropdown, so I wouldn't expect users to provide them with spaces? Or perhaps I am missing something?

File: assets/svelte/functions/Edit.svelte (hunk 13 above)
Comment (davoclavo) [reply assets/svelte/functions/Edit.svelte]: Great idea!

I was also thinking that the editors could be always visible/editable, and no click on the pencil was needed, but happy to keep it as is for now.

File: lib/sequin/functions/minielixir.ex (hunk 12 above)
Comment (davoclavo) [reply lib/sequin/functions/minielixir.ex]: Thanks! I ended up doing a slightly modified version, so the `record` part of the path could also be `action`, `metadata`, or `changes` as well.

```elixir
    unwrapped =
      if function.type == :path do
        [variable | path] = String.split(function.path, ".")

        # NOTE: Using nil context for the variable definition is crucial
        variable = Macro.var(String.to_existing_atom(variable), nil)

        quoted =
          if path == [] do
            variable
          else
            quote do
              get_in(unquote(variable), unquote(path))
            end
          end

        {:ok, quoted}
      else
        function.code |> Code.string_to_quoted!() |> Validator.unwrap()
      end
```

Comment (fazzone) [reply lib/sequin/functions/minielixir.ex]: ahh i thought it would be easy but I didn't test what I had written!  nice

File: lib/sequin_web/live/functions/edit.ex (hunk 11 above)
Comment (davoclavo) [reply lib/sequin_web/live/functions/edit.ex]: Yeah kind of odd and a bit annoying. All of these values are serialized versions of an elixir map, nil, or string; so I had to strip that escaped quote somewhere to do this :(

**Comment by davoclavo** [comment]
I think its ready for merge if nothing else comes up!

Apologies that I mixed a bunch of changes related to `:path` functions in this PR 😅 - I just noticed they were broken while testing the editable fields and ended up in a rabbit hole.

**Comment by RTLS** [comment]
I hit a bug during a quick QA session:

https://www.loom.com/share/02ecb4e41965403c9af17555361651bc

with the stacktrace:

```
[error] [18:14:06.536] GenServer #PID<0.54454.0> terminating
** (CaseClauseError) no case clause matching: %Sequin.Consumers.PathFunction{type: :path, path: "record"}
    (sequin 0.1.0) lib/sequin_web/live/functions/edit.ex:347: SequinWeb.FunctionsLive.Edit.do_encode_test_messages/3
    (sequin 0.1.0) lib/sequin_web/live/functions/edit.ex:307: SequinWeb.FunctionsLive.Edit.do_assign_encoded_messages/2
    (sequin 0.1.0) lib/sequin_web/live/functions/edit.ex:194: SequinWeb.FunctionsLive.Edit.handle_event/3
    (phoenix_live_view 0.20.17) lib/phoenix_live_view/channel.ex:508: anonymous fn/3 in Phoenix.LiveView.Channel.view_handle_event/3
    (telemetry 1.3.0) /Users/carterpedersen/Sequin/sequin/deps/telemetry/src/telemetry.erl:324: :telemetry.span/3
    (phoenix_live_view 0.20.17) lib/phoenix_live_view/channel.ex:260: Phoenix.LiveView.Channel.handle_info/2
    (stdlib 6.2) gen_server.erl:2345: :gen_server.try_handle_info/3
    (stdlib 6.2) gen_server.erl:2433: :gen_server.handle_msg/6
    (stdlib 6.2) proc_lib.erl:329: :proc_lib.init_p_do_apply/3
Last message: %Phoenix.Socket.Message{topic: "lv:phx-GEJRddVIQOiy6f1C", event: "event", payload: %{"event" => "validate", "type" => "hook", "value" => %{"function" => %{"description" => nil, "function" => %{"code" => "", "path" => "record", "type" => "path"}, "id" => nil, "name" => nil}}}, ref: "49", join_ref: "39"} line=2646 pid=<0.54454.0> mfa=:gen_server.error_info/8
```

**Comment by fazzone** [comment]
We were taking a second (third?) look at this again today and actually I think the path function changes aren't really worth it in this PR.  I think longer term we may unify path function with regular minielixir function, but this would more likely take the form of a database migration which converted the path functions into transform functions - but we may not even do that!  For now we can try and keep the path functions as simple as possible and it's okay if they are just an alternate codepath in several places, even if it is a bit awkward.  Sorry for the delay in re-reviewing as well as changing my mind - we do want this merged and we appreciate your patience!

**Comment by davoclavo** [comment]
Quick update: I just removed the path function runner implementation. Other path function-related stuff (like the bug that was mentioned) will be addressed in separate PRs.
